src/common/components/lang/data.ts

```diff
@@ -0,0 +1,368 @@
+export interface Config {
+    name: string
+    nameEn: string
+    isVariant?: boolean // whether the language is one of the variants of the language, default `false`
+    isSource?: boolean // whether the language can be translated from, default `true`
+    isTarget?: boolean // whether the language can be translated to, default `true`
+    direction?: 'ltr' | 'rtl' // direction of the language, default `ltr`
+    rolePrompt?: string // prompt for the role of the translator, default `''`
+    genCommandPrompt?: (sourceLanguageConfig: Required<Config>, quoteStart: string, quoteEnd: string) => string
+    phoneticNotation?: string | false // string for the name of the transcription / transliteration system, `false` if not applicable, default 'transcription'
+    /*
+     `phoneticNotation` is not applicable for languages that have no phonetic transcription, only
+     when the language has neither transcription nor transliteration systems. This can happen
+     only in the below circumstances:
+     1. The language is written in latin alphabets AND using a phonemic orthography, i.e. an
+        orthography (system for writing a language) in which the graphemes (written symbols)
+        correspond to the phonemes (significant spoken sounds) of the language;
+     2. GPT does not support it well;
+     3. This is a variant and only native speakers understand so that it is not necessary to mark
+        its pronunciation.
+    */
+}
+
+interface LanguageConfigs {
+    [key: string]: Config
+}
+
+export const languageConfigs = {
```

Comment:
How about adding type hints here:

```ts
export const languageConfigs: Record<string, Config> = {
```

Reply:
This will make `export type SupportedLanguageCode = keyof typeof languageConfigs` in `lang/lang.ts` not a union type of all supported language codes, but `string`. Haven't find any ways to partially limit only the type of values of Record — there is no available underscore [type placeholder](https://github.com/Microsoft/TypeScript/pull/26349) like in Rust.

![image](https://github.com/openai-translator/openai-translator/assets/20166026/9581259b-f019-482a-bac9-6ff70cbe7fa4)

Expected:

![image](https://github.com/openai-translator/openai-translator/assets/20166026/37fa145d-4eb1-41c1-9f6b-4aaaaeae01d1)

Any ideas on how to fix this?

Reply:
One possible solution is to write an enum. Is this okay?
![IMG_20230523_141506](https://github.com/openai-translator/openai-translator/assets/20166026/037325d0-87d8-4898-ae91-d8358efa4fb8)
![IMG_20230523_141519](https://github.com/openai-translator/openai-translator/assets/20166026/635a5814-137d-42fd-a088-21289ccdcd42)

Reply:
Or, `Array<{ enums, …other_fields }>` with another Map?

Reply:
How about this?

https://www.typescriptlang.org/play?#code/KYOwrgtgBAymAO8D2AnALsAJgGQIYgHMxcDgBnKAbwCgo6oBREAVRigF4oAiUAWla4AaWvQBaACwAS+Cpy4Avcb2kgyXagF9q1YAA9k6KAGMkqtFAA2+IiWABhUwDMAlgVlURdANpxEqDDjWxKRkAHRMrAC6AFwe9PFQILgQwLEA5ACCKSjORviMhBbOZOJpwgn0SSlM6UwERSVQABSsAJRlngnw4qbAaLkAckhouP2m6QCSAAoZHRV0xTBIYChGqVCOuBZkwOX0GnvevgYBeITB5KESKmQxcfNVwDVQaTDOEPBFLlhQduLOIHIwDmD2S6zSgADvQDK8oBaOUA4aYgirFABquBy+DQsTQKDAu068W6vX6RiGIzGIHSUwBAE8AYioAdNEA